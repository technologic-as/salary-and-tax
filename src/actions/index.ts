import Cookie from 'js-cookie';
import { updateIntl } from 'react-intl-redux';
import { change, submit } from 'redux-form';
import { form } from '../components/SalaryParameters';
import { getMessages } from '../translation';
import { Dispatch } from 'redux';
import { SupportedLocale } from '../constants';

export const FORM_CHANGE = 'FORM_CHANGE';

export const formChange = (values: any) => ({ type: FORM_CHANGE, ...values });
export type FormChange = { type?: string; [k: string]: unknown };

export const inputHasChanged =
  (values: { [k: string]: unknown } = {}) =>
  (dispatch: Dispatch) => {
    if (values.locale) {
      switchLanguage(values.locale as SupportedLocale)(dispatch);
    }
    return dispatch(formChange(values));
  };

export const switchLanguage = (locale: SupportedLocale) => (dispatch: Dispatch) => {
  Cookie.set('locale', locale);
  dispatch(updateIntl({ locale: locale, messages: getMessages(locale) }));
};

export const dividendsClick = (point: { x: number }) => (dispatch: Dispatch) => {
  dispatch(change(form, 'surplus.include', true));
  dispatch(change(form, 'surplus.amount', point.x, true));
  dispatch(submit(form));
};
