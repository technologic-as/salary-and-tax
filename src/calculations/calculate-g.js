import round from './round';

export const calculateG = g => round(99858 * g);

export default calculateG;
