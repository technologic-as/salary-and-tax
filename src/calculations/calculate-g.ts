import round from './round';
import { taxConstants } from '../constants';

export const calculateG = (g: number) => round(taxConstants.oneG * g);

export default calculateG;
