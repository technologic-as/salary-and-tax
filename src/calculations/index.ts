import { defaultSalaryParameters, getSalaryCalculations, SalaryCalculationsState } from './salary';
import { DividendsTaxCalculation, getDividendsTaxCalculations, getTaxCalculations, TaxCalculations } from './tax';
import { ParamProps } from '../constants';

export { defaultSalaryParameters } from './salary';

type Calculations = {
  dividends: DividendsTaxCalculation;
  tax: TaxCalculations;
  afterTotal: number;
  salary: SalaryCalculationsState;
};

export const calculate = (parameters: ParamProps = defaultSalaryParameters): Calculations => {
  const salary = getSalaryCalculations(parameters);
  const tax = getTaxCalculations(salary.income);
  const dividends = getDividendsTaxCalculations(salary.surplus);
  const afterTotal = tax.afterTax + dividends.afterDividendsTax;

  return { salary, tax, dividends, afterTotal };
};

export default calculate;
