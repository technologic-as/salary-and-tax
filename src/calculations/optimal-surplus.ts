import { calculateG } from './calculate-g';
import { calculate } from './index';
import { defaultSalaryParameters } from './salary';
import { ParamProps } from '../constants';

export type OptimalSurplus = {
  total: number;
  surplus: any;
  salaryBasis: number;
  dividends: number;
  netIncome: number;
  grossIncome: number;
  salaryAboveSevenG: boolean;
};

export const optimalSurplus = (parameters: ParamProps = defaultSalaryParameters): OptimalSurplus[] => {
  const initialCalculation = calculate(parameters);
  const companyIncome = initialCalculation.salary.companyIncome;
  const increments = parameters.graph.increments;
  const steps = Math.round(companyIncome / increments);

  const sevenG = calculateG(7.1);

  return Array.from(Array(steps).keys())
    .map((i: number) => i * increments)
    .map((amount: number) =>
      calculate({
        ...parameters,
        surplus: {
          amount,
          include: true,
        },
      })
    )
    .map(
      ({
        afterTotal,
        dividends: { afterDividendsTax, surplus },
        salary: { withoutPension, withoutSurplus },
        tax: { afterTax },
      }) => ({
        total: afterTotal,
        surplus,
        dividends: afterDividendsTax,
        grossIncome: withoutPension,
        netIncome: afterTax,
        salaryBasis: withoutSurplus,
        salaryAboveSevenG: withoutPension > sevenG,
      })
    );
};

export default optimalSurplus;
