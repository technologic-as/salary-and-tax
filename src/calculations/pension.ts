import { calculateG } from './calculate-g';
import round from './round';
import { PensionProps, PensionStep } from '../constants';

export const pensionConstants: {
  pension: PensionProps;
} = {
  pension: {
    include: true,
    step1: {
      rate: 2,
      threshold: { low: calculateG(1), high: calculateG(7.1) },
    },
    step2: {
      rate: 2,
      threshold: { low: calculateG(7.1), high: calculateG(12) },
    },
  },
};

const calculateStep = (income: number, { rate, threshold: { low, high } }: PensionStep, include: boolean): number => {
  if (!include) {
    return 0;
  }
  const maxStep = high - low;
  const aboveThreshold = income - low;
  const pension = (Math.min(maxStep, aboveThreshold) * rate) / 100;

  return round(aboveThreshold > 0 ? pension : 0);
};

export const calculateStep2 = (salary: number, { step2, include }: PensionProps) => {
  return calculateStep(salary, step2, include);
};

export const calculateStep1 = (salary: number, { step1, include }: PensionProps) => {
  return calculateStep(salary, step1, include);
};

type PensionCalculation = {
  pension: number;
  step1: number;
  step2: number;
  after: number;
};

export const getPension = (salary: number, { pension }: { pension: PensionProps }): PensionCalculation => {
  const step1 = calculateStep1(salary, pension);
  const step2 = calculateStep2(salary, pension);
  const result = pension.include ? round(step1 + step2) : 0;

  return { pension: result, step1, step2, after: salary - result };
};

export default getPension;
