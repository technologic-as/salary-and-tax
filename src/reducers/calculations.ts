import { FORM_CHANGE, FormChange } from '../actions';
import { calculate, defaultSalaryParameters } from '../calculations';
import { DividendsTaxCalculation, TaxCalculations } from '../calculations/tax';
import { SalaryCalculationsState } from '../calculations/salary';
import { ParamProps } from '../constants';

const initialState: CalculationsState = {
  ...calculate(defaultSalaryParameters),
};

export type CalculationsState = {
  dividends: DividendsTaxCalculation;
  tax: TaxCalculations;
  afterTotal: number;
  salary: SalaryCalculationsState;
};

export const calculationsReducer = (
  state: CalculationsState = initialState,
  action: FormChange = {},
  { parameters }: { parameters?: ParamProps } = {}
): CalculationsState => {
  switch (action.type) {
    case FORM_CHANGE:
      return Object.assign({}, state, {
        ...calculate(parameters),
      });
    default:
      return state;
  }
};

export default calculationsReducer;
