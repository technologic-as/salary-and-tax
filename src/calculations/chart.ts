import { OptimalSurplus, optimalSurplus } from './optimal-surplus';
import { defaultSalaryParameters } from './salary';
import { ParamProps } from '../constants';

export type ChartState = {
  total: (any | number)[][];
  dividends: (any | number)[][];
  netIncome: (any | number)[][];
  grossIncome: (any | number)[][];

  maxTotal: any[][];
  surplus: any[][];
  sevenG: number[][];
};

export const getChart = (parameters: ParamProps = defaultSalaryParameters): ChartState => {
  const data = optimalSurplus(parameters);
  const dividends = data.map((i) => [i.surplus, i.dividends]);
  const surplus = data.map((i) => [i.surplus, i.surplus]);
  const grossIncome = data.map((i) => [i.surplus, i.grossIncome]);
  const netIncome = data.map((i) => [i.surplus, i.netIncome]);
  const total = data.map((i) => [i.surplus, i.total]);
  const closeToSeven = data
    .filter((i) => i.salaryAboveSevenG)
    .reduce((a: OptimalSurplus, b) => (a.salaryBasis < b.salaryBasis ? a : b));
  const sevenG = [[closeToSeven.surplus, closeToSeven.grossIncome]];
  const max = data.reduce((a: OptimalSurplus, b) => (a.total > b.total ? a : b));
  const maxTotal = [[max.surplus, max.total]];

  return {
    sevenG,
    maxTotal,
    grossIncome,
    netIncome,
    surplus,
    dividends,
    total,
  };
};

export default getChart;
