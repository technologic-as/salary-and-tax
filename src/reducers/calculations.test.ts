import { FORM_CHANGE } from '../actions';
import * as calculations from '../calculations';
import reducer from './calculations';

describe('calculations', () => {
  it('should have initial state', function () {
    expect(reducer()).toMatchSnapshot();
  });
  describe('FORM_CHANGE', () => {
    it('should recalculate with parameters', () => {
      const spy = jest.spyOn(calculations, 'calculate');
      const parameters = calculations.defaultSalaryParameters;
      reducer(undefined, { type: FORM_CHANGE }, { parameters });
      expect(spy).toHaveBeenCalledWith(parameters);
    });
  });
});
