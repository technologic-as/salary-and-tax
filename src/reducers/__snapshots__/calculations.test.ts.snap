// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`calculations should have initial state 1`] = `
Object {
  "afterTotal": 842620,
  "dividends": Object {
    "afterDividendsTax": 0,
    "dividends": 0,
    "dividendsTax": 0,
    "surplus": 0,
    "surplusTax": 0,
    "upwardsAdjustment": 0,
  },
  "salary": Object {
    "companyIncome": 1712700,
    "employerFee": 211648,
    "income": 1315701,
    "pension": 24524,
    "surplus": 0,
    "theCut": 190300,
    "totalCuts": 587299,
    "turnover": 1903000,
    "vacationSavings": 160827,
    "withoutEmployerFee": 1501052,
    "withoutPension": 1315701,
    "withoutSurplus": 1712700,
    "withoutVacationSavings": 1340225,
  },
  "tax": Object {
    "afterTax": 842620,
    "commonIncome": 1141101,
    "income": 1315701,
    "incomeTax": 251042,
    "minimumDeduction": 109950,
    "personalAllowance": 64650,
    "socialSecurityDeduction": 105256,
    "step1": 1318,
    "step2": 15036,
    "step3": 43603,
    "step4": 56826,
    "step5": 0,
    "totalTax": 473081,
  },
}
`;
